Description: Do not allow devices in an offscreen hierarchy to take grabs.
Author: Robert Carr <racarr@canonical.com>
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=658563
Bug-Ubuntu: https://launchpad.net/bugs/804009

Index: gtk+-3.20.5/gtk/gtkmain.c
===================================================================
--- gtk+-3.20.5.orig/gtk/gtkmain.c	2016-05-20 16:26:01.851409723 +0200
+++ gtk+-3.20.5/gtk/gtkmain.c	2016-05-20 16:26:01.847409748 +0200
@@ -2284,9 +2284,14 @@
 {
   GtkWindowGroup *group;
   GtkWidget *old_grab_widget;
+  GdkWindow *toplevel;
 
   g_return_if_fail (GTK_IS_WIDGET (widget));
   g_return_if_fail (GDK_IS_DEVICE (device));
+  
+  toplevel = gdk_window_get_toplevel (gtk_widget_get_window (widget));
+  if (toplevel && gdk_window_get_window_type (toplevel) == GDK_WINDOW_OFFSCREEN)
+    return;
 
   group = gtk_main_get_window_group (widget);
   old_grab_widget = gtk_window_group_get_current_device_grab (group, device);
